CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,288 @@
+    if (SUCCEEDED(hr))
+    {
+        if (s_aumid != nullptr)
+        {
+            hr = toastStatics->CreateToastNotifierWithId(HStringReference(s_aumid).Get(), notifier);
+        }
+
+        else
+        {
+            hr = toastStatics->CreateToastNotifier(notifier);
+        }
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::CreateToastNotification(ABI::Windows::Data::Xml::Dom::IXmlDocument *content, IToastNotification **notification)
+{
+    ComPtr<IToastNotificationFactory> factory;
+    auto hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotification).Get(),
+        &factory);
+    if (SUCCEEDED(hr))
+    {
+        hr = factory->CreateToastNotification(content, notification);
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::get_History(DesktopNotificationHistoryCompat *value)
+{
+    EnsureRegistered();
+
+    ComPtr<IToastNotificationManagerStatics> toastStatics;
+    HRESULT hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotificationManager).Get(),
+        &toastStatics);
+    if (SUCCEEDED(hr))
+    {
+        ComPtr<IToastNotificationManagerStatics2> toastStatics2;
+        hr = toastStatics.As(&toastStatics2);
+        if (SUCCEEDED(hr))
+        {
+            ComPtr<IToastNotificationHistory> history;
+            hr = toastStatics2->get_History(&history);
+            if (SUCCEEDED(hr))
+            {
+                *value = DesktopNotificationHistoryCompat(s_aumid, history);
+            }
+        }
+    }
+
+    return S_OK;
+}
+
+bool DesktopNotificationManagerCompat::CanUseHttpImages()
+{
+    return IsRunningAsUwp();
+}
+
+void DesktopNotificationManagerCompat::EnsureRegistered()
+{
+    // If not registered AUMID yet
+    if (!s_registeredAumidAndComServer)
+    {
+        // Check if Desktop Bridge
+        if (IsRunningAsUwp())
+        {
+            // Implicitly registered, all good!
+            s_registeredAumidAndComServer = true;
+        }
+        else
+        {
+            // Otherwise, incorrect usage
+            throw std::exception("You must call RegisterAumidAndComServer first.");
+        }
+    }
+
+    // If not registered activator yet
+    if (!s_registeredActivator)
+    {
+        // Incorrect usage
+        throw std::exception("You must call RegisterActivator first.");
+    }
+}
+
+bool DesktopNotificationManagerCompat::IsRunningAsUwp()
+{
+    if (!s_hasCheckedIsRunningAsUwp)
+    {
+        s_hasCheckedIsRunningAsUwp = true;
+
+        // https://stackoverflow.com/questions/39609643/determine-if-c-application-is-running-as-a-uwp-app-in-desktop-bridge-project
+        UINT32 length;
+        wchar_t packageFamilyName[PACKAGE_FAMILY_NAME_MAX_LENGTH + 1];
+        LONG result = GetPackageFamilyName(GetCurrentProcess(), &length, packageFamilyName);
+        s_isRunningAsUwp = result == ERROR_SUCCESS;
```

Comment:
Are these methods supposed to be threadsafe? This can return false for thread2 while thread1 is calling GetPackageFamilyName

Reply:
Ah really good catch! Moving setting s_hasCheckedIsRunningAsUwp down to the end AFTER assigning the s_isRunningAsUwp should fix that, thanks!